framework/Optimizers/GradientBasedOptimizer.py

```diff
@@ -265,86 +267,124 @@ def checkConvergence(self):
-        self.raiseAMessage("Grad Norm : "+"%8.2E"% (gradNorm)+" | Relative Diff: "+"%8.2E"% (relativeDifference)+" | Abs Diff     : "+"%8.2E"% (absDifference)+" |")
-        self.raiseAMessage("Step Size : "+"%8.2E"% (lastStep))
-        self.raiseAMessage("Input Location :" +str(varK))
-        ## set up status going forward
-        # if new point is better, accept it and move forward
-        if newerIsBetter:
-          self.status[traj]['reason'] = 'found new opt point'
-          self.raiseADebug('Accepting potential opt point for improved loss value')
-          #TODO REWORK this belongs in the base class optimizer; grad shouldn't know about multilevel!!
-          #  -> this parameter is how multilevel knows that a successful perturbation of an outer loop has been performed
-          self.mlActiveSpaceSteps[traj] += 1
-          converged = minStepSizeCheck or sameCoordinateCheck or gradientNormCheck or absoluteTolCheck or relativeTolCheck
-        # if newer point is not better, we're keeping the old point, and sameCoordinate, absoluteTol, and relativeTol aren't applicable
-        else:
-          self.status[traj]['reason'] = 'rejecting bad opt point'
-          self.raiseADebug('Rejecting potential opt point for worse loss value: "{}" vs "{}"'.format(oldVal,currentLossValue))
-          # cut the next step size to hopefully stay in the valley instead of climb up the other side
-          self.recommendToGain[traj] = 'cut'
-          converged = gradientNormCheck or minStepSizeCheck
-        if converged:
-          reasons = []
-          if sameCoordinateCheck:
-            reasons.append("same coordinate")
-          if gradientNormCheck:
-            reasons.append("gradient norm")
-          if absoluteTolCheck:
-            reasons.append("absolute tolerance")
-          if relativeTolCheck:
-            reasons.append("relative tolerance")
-          if minStepSizeCheck:
-            reasons.append("minimum step size")
-          self.raiseAMessage("Trajectory: "+"%8i"% (traj) +"   converged. Reasons: "+', '.join(reasons))
-          self.convergeTraj[traj] = True
-          self.removeConvergedTrajectory(traj)
+      else:
+        gradNorm = np.nan
+        gradientNormCheck = False
+      printProgress('Gradient magnitude',gradientNormCheck,gradNorm,self.gradientNormTolerance)
+      converged = converged or gradientNormCheck
+
+      # if accepting new point, then "same coordinate" and "abs" and "rel" checks are also valid reasons to converge
+      if newerIsBetter:
+        #absolute tolerance
+        absLossDiff = abs(currentLossVal-oldLossVal)
+        self.convergenceProgress[traj]['abs'] = absLossDiff
+        absTolCheck = absLossDiff <= self.absConvergenceTol
+        printProgress('Absolute Loss Diff',absTolCheck,absLossDiff,self.absConvergenceTol)
+        converged = converged or absTolCheck
```

Comment:
The checks are quite inexpensive, and we offer them as outputs on each iteration to the user, so it is worth checking them all, in my opinion.

Originally, the "or" check was done all in one line, but this was not conducive to future development involving addition of other convergence criteria. In the way it is now, a developer need only copy one of the existing check blocks.